RFC: add `(strict_includes)` in `(library)` and `(executable)`

hnrgrgr opened the pull request:
Sometimes, in order to enforce abstractions, we might want to build a library without allowing access to all the `.cmi` of the "recursive" libraries it depends upon. For instance, we might want to call the OCaml compiler with only the `-I` of the explicitly listed libraries and not the `-I` of their recursives closures.

This patch allows this behaviour by introducing an optional `(strict_includes)` stanza in `(library)` and `(executables)`.

The default behaviour is not changed and obviously this patch preserves the recursive closure of dependent libraries when calling the executable linkers.

diml commented:
Thanks, that seems like a reasonable feature.

A few remarks before looking at the code:
- I don't really like talking about include directories in jbuild files, it feels a bit low-level. Maybe we could have `(strict_libraries (foo bar))`  instead of `(libraries (foo bar))` to get the strict behavior
- Does the `(strict_include)` has an affect in the following jbuild file:
```scheme
(library
 ((name a)
  (modules (a))))

(library
 ((name b)
  (modules (b))
  (libraries (a))))

(library
 ((name c)
  (modules (c))
  (libraries (b))))

(library
 ((name d)
  (modules (d))
  (strict_includes)
  (libraries (c))))
```
? Given that all the cmi will end up in the same directory, I assume no, which is not good. #427 would solve this issue, would you consider implementing it? :)

hnrgrgr commented:
>I don't really like talking about include directories in jbuild files

Agreed.

> Given that all the cmi will end up in the same directory, I assume no, which is not good. #427 would solve this issue, would you consider implementing it? :)

You are right. I will have a look into #427, but probably not before the end of the week.

hnrgrgr committed:
WIP #427

hnrgrgr committed:
WIP: strict inclusion of library

Sometimes, in order to enforce abstractions, we might want to build a
library without allowing access to all the `.cmi` of the 'recursive'
libraries it depend upon. For instance, we might want to call the
OCaml compiler with only the `-I` of the explicitly listed libraries
and not the `-I` of their recursives closures.

This patch allow this behaviour by introducing an optional
`!` as first character of a dependent library.

The default behaviour is not changed and obviously this patch
preserves the recursive closure of dependent libraries when calling
the executable linkers.

hnrgrgr committed:
WIP: compatibility layer

hnrgrgr commented:
I found some time earlier than expected for prototyping #427, I would appreciate your feedback.

The current patch only moves the OCaml objects (`Foo_a.{cmi,cmo,cmx,o}`) into `.foo.objs`, the library files (`foo.{cma,cmxa,a}`) and others stubs are still in their current place.

Regarding the 'strict' inclusion of library, I tried to gave it more thoughs and the disjunction between `(libraries)` and `(strict_libraries)` puzzles me and I am not sure of the best user interface. Just for quick prototyping I implemented a third solution `(libraries (a !b c))` where `a` and `c` behaves as usual and onlye the library `b` has the 'strict' attributes. The same syntax is also used in `lib.requires.sexp`, this allows the 'strict' attribute to be propagated (I don't have a specific usage in mind but it seems somewhat more coherent).

diml commented:
> I found some time earlier than expected for prototyping #427, I would appreciate your feedback.

Cool, thanks. I'd rather not rely on symlinks as on Windows this will cause one more file copy. Does it work if we just change the compiler invocation to:

```
$ ocamlc ... -o src/.foo.objs/blah.cmo
```

> The current patch only moves the OCaml objects (Foo_a.{cmi,cmo,cmx,o}) into .foo.objs, the library files (foo.{cma,cmxa,a}) and others stubs are still in their current place.

That should be fine.

> Regarding the 'strict' inclusion of library, I tried to gave it more thoughs and the disjunction between (libraries) and (strict_libraries) puzzles me and I am not sure of the best user interface. Just for quick prototyping I implemented a third solution (libraries (a !b c)) where a and c behaves as usual and onlye the library b has the 'strict' attributes. The same syntax is also used in lib.requires.sexp, this allows the 'strict' attribute to be propagated (I don't have a specific usage in mind but it seems somewhat more coherent).

So, how does it work, do libraries essentially have two sets of dependencies? API ones and link time ones? Note that this is actually what we do inside Jane Street, however we do it automatically: the API dependencies are computed by reading the output of ocamlobjinfo. Though this requires being able to associate modules to libraries.

samoht commented:
> do libraries essentially have two sets of dependencies? API ones and link time ones

I think this would be a sane default: you can only use symbols in the libraries defined in the `libraries` field. I am not even sure we need to keep the current behaviour at all.

hnrgrgr commented:
> I'd rather not rely on symlinks as on Windows this will cause one more file copy. Does it work if we just change the compiler invocation to: `$ ocamlc ... -o src/.foo.objs/blah.cmo`

That's what the patch does, it currently invokes the compiler the way you proposed.

The symlink is the other way round (e.g. `src/blah.cmo` references `src/.foo.objs/blah.cmo`). and it is not used by default. I added the symlink in a second time, just to preserves the compatibility with invocation like `jbuilder build src/blah.cmo`, which are used in the blockbox test (easily fixable) but also in external project (like Janestreet's base). This symlink things is only implemented in the third patch of the series and can be easily removed.

hnrgrgr commented:
> >  do libraries essentially have two sets of dependencies? API ones and link time ones

Sort of, the API one being included in the link time one. In term of implementation, I just added a boolean in `Lib.t`, stating whether the libraries should be in the 'include path' or not.

```ocaml
type t =
  | Internal of Internal.t * bool
  | External of Findlib.package * bool
```

> I think this would be a sane default: you can only use symbols in the libraries defined in the libraries field. I am not even sure we need to keep the current behaviour at all.

That makes sense, and would simplify the implementation.

diml commented:
> That's what the patch does, it currently invokes the compiler the way you proposed.

> The symlink is the other way round (e.g. src/blah.cmo references src/.foo.objs/blah.cmo). and it is not used by default. I added the symlink in a second time, just to preserves the compatibility with invocation like jbuilder build src/blah.cmo, which are used in the blockbox test (easily fixable) but also in external project (like Janestreet's base). This symlink things is only implemented in the third patch of the series and can be easily removed.

Ah, I see. Sorry, I didn't read the patch carefully enough. That makes sense.

diml commented:
I'm a bit lost in all the discussions, can someone post a jbuild file defining several libraries with some use of the new syntax and explain what libraries are expected to be seen by each library?

hnrgrgr commented:
Here is the current semantics and syntax:

```
(library
 ((name a)
  (libraries (lwt))
  (modules (a))))

;; Default behaviour (unchanged)
(library
 ((name b)
  (modules (b)) ;; Sees `a`, `lwt` (and its dependecies)
  (libraries (a))))

;; Importing `b` without visibility on its dependencies
(library
 ((name c)
  (modules (c)) ;; Only sees `b`
  (libraries (!b))))

;; By default, importing `c` makes `b` visible, but not `a` nor `lwt`...
(library
 ((name d)
  (modules (d)) ;; Only sees `c`
  (libraries (c))))

;; Explicitly importing `b` allows to make `a` and `lwt` visible...
(library
 ((name e)
  (modules (e)) ;; Sees `a`, `b`, `c`, `d, `lwt`,...
  (libraries (b d))))
```

diml commented:
OK, thanks. BTW are there cases where we want the strict behavior only for a subset of the dependencies?

About changing the default, you can try but I'm pretty sure that this will break tons of stuff.

diml commented:
Wait, in your example, shouldn't `d` see `b` as well?

hnrgrgr commented:
> BTW are there cases where we want the strict behavior only for a subset of the dependencies?

I am not sure. The patch was simpler this way and I am not against using `(libraries)` and `(strict_libraries)`, I just choosed the simpler syntax for quick prototyping.

Thanks for your feedbacks.

hnrgrgr commented:
> Wait, in your example, shouldn't d see b as well?

It does, I fixed the comment.

diml commented:
OK, so in the new world, is writing `!x` good and writing `x` bad? For instance writing `!core` is never a good idea as you would need to write `!core !core_kernel !base` even if you only refer to `Core` in your code, which is weird.

diml commented:
For instance, I was trying to think of this feature as:

- `!x` is for dependencies used only in .mli files
- `x` is for all other dependencies

but that still doesn't work for Core.

diml commented:
BTW, I guess Core is not the common case. If it help with real use cases, I'm fine with the feature as it is now.

diml commented:
I thought a bit more about this, it seems to me that:

- the strict mode is better in general
- it is only an issue for libraries that extend other ones

Based on that I suggest:

1. to make the strict behavior a setting that one might set globally for a project, this could fit in the new environment feature (#419)
2. to allow libraries to express that they extend another library, so that for instance adding `core` to the list of dependencies is the same as adding `core base core_kernel`, even when the strict behavior is on

Then in a distant future we could try to make the strict behavior the default.

What do you think?

diml commented:
@hnrgrgr, could you extract the implementation of #427 into a separate PR? This will allow to make progress on some other PRs until a consensus has been reached for this one.

hnrgrgr commented:
@diml Rebased the patch and made anew PR, see #472.

diml commented:
Great, thanks!

diml commented:
I was thinking about this again and it seems to me that the best would be a compiler feature to express that we want to reexport a module. i.e. something like this:

```ocaml
module X = Foo.X [@@reexport]
include Y [@@reexport]
```

that would effectively inline the cmi contents instead of just creating aliases

diml commented:
I found several cases where I wanted this feature. Let's go ahead with it without bothering with libraries that do complicated stuff for now. Now that we have environments, we can easily setup this property globally for a project.

I suggest to add a field `(transitive_deps <mode>)` where `<mode>` is either `visible` or `hidden`. This field should be allowed in `library`, `executable`, ... and `env` stanzas. @hnrgrgr, are you happy to look at this?

Octachron commented:
Isn't there also a problem with any libraries that defines a type abbreviation to one of its dependencies?
For instance, with three files:
```OCaml
(* c.mli *)
type t = A
```
```OCaml
(* b.mli *)
type t = C.t
```
```OCaml
(* a.ml *)
type _ t = X: int t
let f: B.t t -> 'a = function _ -> .
```
The last file `a.ml` is well typed if and only if `a.cmi` is visible in scope.

rgrinberg commented:
@Octachron doesn't that make sense? If a client code is relying on some library to introduce some type equalities (or `A.t` not being abstract in this case - if I'm understanding this right), then they should introduce a dependency on the library of `a.ml` explicitly.

Octachron commented:
I imagine that asking users to do some dependency analysis on their own in those corner cases is sensible, but this is also a warning that there is observable differences between building with and without the transitive closure of dependencies. Moreover, hiding type equalities do not always result in errors: Hiding an equality `t=float` can impede some of the adhoc float optimisation for instance.

diml commented:
Technically, it doesn't have to break the optimization right? I mean, if a type `Y.t` is declared as `type t = X.t` and you know that `X.t` is a float, you could record that fact when compining `Y`.

Overall, I feel like they will be corner cases, however we should be able to solve them. Providing this feature and using it will allow us to discover them.

Octachron commented:
Sure, the cmi representation could be changed to record the full type declaration and hide it, rather than link to the original definition like currently. I rather agree with your analysis, at most it could be nice to document those known corner cases.

diml commented:
Documenting them seems right

rgrinberg commented:
@hnrgrgr if you don't have time, we can help out with rebasing this work as well. Just let us know.

rgrinberg commented:
@diml I'm wondering if the project itself is actually the correct place to set this mode. I assume that we'd start the transition to this new mode, and eventually, we'd make this the default. In which case this option would be obsolete. At the moment, I'm fairly convinced that the current recursive behavior is useless apart from backwards compatibility concerns. So I think it would be more trouble than its worth to make an option that would eventually removed available in so many places.

Extending libraries can be handled with a field like `exported_libraries`. This field will default to the empty list, but would be set for `core_kernel` in `core` for example. I think that there would be no need to set `base` here, as I'd expect that `core_kernel` would already export it. I'd expect that this particular field would indeed be recursive. The only issue I see is that we'd only make it possible to define external grouped libraries like this in dune. That seems acceptable.

> For instance, I was trying to think of this feature as:
>
> - `!x` is for dependencies used only in .mli files
> - `x` is for all other dependencies
>
> but that still doesn't work for Core.

This concern seems orthogonal to me. Nevertheless, it seems useful to be able to control interface dependencies this way. How about adding an ordered set valued `interface_libraries` field?. `:standard` in this case mean would mean to use the set of dependencies in the `libraries` field unioned with the closure of `exported_libraries`. In any case, we don't have to do this level of control right away.

diml commented:
It would indeed be nice to make it the default in `(lang dune 2.0)`. However, I don't know yet how realistic that is, we might find more issues on the way. It seems to me that the most important is to able to enable it for a given project. We shouldn't worry about core_kernel, I expect that we should be able to solve the problem for core_kernel without adding anything in dune.

That is how I expect things to go:
- we add support only for the `transitive_deps` field (we drop all the other ideas proposed in this PR)
- projects that want to try it add `(env (_ (transitive_deps hidden)))` in their toplevel `dune` file
- as expected, we realize that it is a better default and all the actively developed projects naturally migrate to it

Then when we are ready to release `(lang dune 2.0)` we have the choice between the following:
- if at this point it is obvious that `(transitive_deps hidden)` is always better and there are no more cases where it cannot be used, we simply drop this field in `(lang dune 2.X)` and always hide transitive dependencies for projects using `(lang dune 2.X)`
- if it's clear that it is a better default but there are still cases where we can't use `(transitive_deps hidden)`, we simply flip the default from `visible` to `hidden`
- if it's not clear that it is a better default, we change nothing

rgrinberg commented:
Alright, that sounds like a good idea. Btw, I think that rebasing this PR is not realistic - far too much has changed. I guess I could try writing this myself if Gregoire is preoccupied.

hnrgrgr commented:
I also agree with the proposal. I won't be able to work on it before
October. Feel free to rebase/rewrite, if it is more urgent.

On Mon, Sep 3, 2018 at 2:10 PM Rudi Grinberg <notifications@github.com>
wrote:

> Alright, that sounds like a good idea. Btw, I think that rebasing this PR
> is not realistic - far too much has changed. I guess I could try writing
> this myself if Gregoire is preoccupied.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ocaml/dune/pull/430#issuecomment-418096792>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAhgK8E1RVizxvQI5M2GX-XNLX-tYBuEks5uXRwggaJpZM4Rl03f>
> .
>
-- 

Grégoire

rgrinberg commented:
I think it's unlikely that I'd get to work on this before the ICFP as well. Feel free to take your time.

lpw25 commented:
Sorry for coming to this very late, but I'm not sure this is the right approach to this issue. Basically, there are two kinds of lookups in the type environment:

1. Looking up the path and definition associated with a particular string.
2. Looking up a definition based on a given path.

For example, when you write:
```ocaml
type t = Foo.t
```
in a file `bar.mli` you will do a lookup of the first kind for the string `Foo` when compiling `bar.mli` whereas users of `bar.cmi` my well do a lookup of the second kind for the path `Foo`.

Transitive dependencies are a bad thing for lookups of the first kind: you only want the user to be able to write `Foo` in their file if they actually specified the library containing `Foo` as a dependency. 

However, for lookups of the second kind the compiler always assumes that the transitive closure of dependencies will be available -- and which lookups it does is essentially an implementation detail that the user has no real control over.

(A third kind of lookup is looking up a `.cmx` file for a path -- but this is really just a variation on the second kind of lookup.)

So if you provide the transitive dependencies you get extra names in scope, whereas if you don't you get unexpected errors.

With the compiler as it is now the only known workaround for this issue is to have the build system `-open` a module that shadows the `.cmi` files which should be visible as paths but not as strings. This is an ugly hack and doesn't give great error messages.

The best solution would be to change how the compiler works. All that is needed is a distinction between include paths that should be available for the both kinds of lookup and include paths that should only be available for the second kind. I think that this would be fairly easy to add now thanks to @diml's recent changes to how the initial environment is represented.

rgrinberg commented:
I revived this in #1745
